walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,53 +15,122 @@
+     * optionalImportantRanges - store time ranges that are important for mandatory as well as optional attendees 
+     * busyIntervals - eventualy store merged important time ranges
+     * optionalBusyRanges - same as above but for mandatory and optional attendees
+     * freeIntervals - used to collect intevals that are between busyIntervals
+     * freeRangesWithOptionals - same as above but for mandatory and optional attendees   
      */
     ArrayList<TimeRange> importantRanges = new ArrayList<TimeRange>();
-    Stack<TimeRange> busyIntervals = new Stack<TimeRange>();
-    ArrayList<TimeRange> freeIntervals = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> optionalImportantRanges = new ArrayList<TimeRange>();
+    Stack<TimeRange> busyRanges = new Stack<TimeRange>();
+    Stack<TimeRange> optionalBusyRanges = new Stack<TimeRange>();
+    ArrayList<TimeRange> freeRanges = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> freeRangesWithOptionals = new ArrayList<TimeRange>();
     
+    //list that helps in optimizing search for important events
+    LinkedList<Event> eventsList = new LinkedList<Event>(events);
+    LinkedList<Event> eventsLeft = eventsList;
+    //populates importantRanges and optionalImportantRanges with ranges that are important for mandatory attendees
     for(String attendee : request.getAttendees()){
-      for (Event event : events){
+      eventsList = eventsLeft;
+      eventsLeft = new LinkedList<Event>();
+      for (Event event : eventsList){
         if(event.getAttendees().contains(attendee)){
           importantRanges.add(event.getWhen());
+          optionalImportantRanges.add(event.getWhen());
+        } else{
+          eventsLeft.add(event);
         }
       }
     }
+    //same as above but for optional attendees
+    eventsList = new LinkedList<Event>(events);
+    eventsLeft = eventsList;
+    for(String attendee : request.getOptionalAttendees()){
+      eventsList = eventsLeft;
+      eventsLeft = new LinkedList<Event>();
+      for (Event event : eventsList){
+        if(event.getAttendees().contains(attendee)){
+          optionalImportantRanges.add(event.getWhen());
+        } else{
+          eventsLeft.add(event);
+        }
+      }
+    }
+    // after two above steps in optionalImportantRanges we have important ranges for optional and mandatory attendees and in
+    // importantRanges ranges important for only mandatory attendees
+
+    // if there are no occupied or optionaly occupied intervals to consider we can immediately return list with WHOLE_DAY
+    // or empty list if WHOLE_DAY is not enaugh to meet duration requirement
+    if(optionalImportantRanges.size() == 0){
+      if(request.getDuration() <= TimeRange.WHOLE_DAY.duration()){
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      } else{ 
+        return Arrays.asList();
+      }
+    }
```

Comment:
Will this part be covered naturally later (line 125-129)? If so, please avoid creating edge cases.